api/provisioning/v1alpha1/provisioningrequest_webhook.go

```diff
@@ -64,3 +95,59 @@ func (r *ProvisioningRequest) ValidateDelete() (admission.Warnings, error) {
 	// TODO(user): fill in your validation logic upon object deletion.
 	return nil, nil
 }
+
+func (r *ProvisioningRequest) validateCreateOrUpdate(oldPr *ProvisioningRequest) error {
+	clusterTemplate, err := r.GetClusterTemplateRef(context.TODO(), webhookClient)
+	if err != nil {
+		return err
+	}
+
+	if err = r.ValidateTemplateInputMatchesSchema(clusterTemplate); err != nil {
+		return err
+	}
+
+	// We only validate the ClusterInstance input here, not the PolicyTemplate input since
+	// its schema is not just for ProvisioningRequest.
+	newPrClusterInstanceInput, err := r.ValidateClusterInstanceInputMatchesSchema(clusterTemplate)
+	if err != nil {
+		return err
+	}
+
+	if oldPr == nil {
+		// ProvisioningRequest is being created, no immutable fields to check
+		return nil
+	}
+
+	// Check for updates to immutable fields in the ClusterInstance input.
+	// Once provisioning has started or reached a final state (Completed or Failed),
+	// updates to immutable fields in the ClusterInstance input are disallowed,
+	// with the exception of scaling up/down when Cluster provisioning is completed.
+	// TODO: consider exposing the conditions in the API.
+	crProvisionedCond := meta.FindStatusCondition(r.Status.Conditions, "ClusterProvisioned")
+	if crProvisionedCond != nil && crProvisionedCond.Reason != "Unknown" {
```

Comment:
I think we should make these conditions constants and add a comment that they should be kept in sync with those under [internal/controllers/utils/conditions.go.](https://github.com/openshift-kni/oran-o2ims/blob/main/internal/controllers/utils/conditions.go#L22-L46)
It would be easy to get the conditions mismatched.
It would actually be even better to move these conditions to completely into the `api/provisioning` directory and have them referenced in both the webhook and the controller code under `internal/`.

Reply:
> It would actually be even better to move these conditions completely into the api/provisioning directory and have them referenced in both the webhook and the controller code under internal/.

Yep, I was thinking the same, since the movement may require a lot of places to be updated so I added the TODO for that. Maybe it's not clear enough.

Reply:
Ok, I see it now. Sounds good to me.